cpp/src/fil/fil.cu

```diff
@@ -374,14 +407,14 @@ struct dense_forest<dense_node<real_t>> : forest {
        roots of all trees (node 2), and so on.
     */
     int global_node = 0;
-    for (int tree = 0; tree < num_trees_; ++tree) {
+    for (int tree = 0; tree < this->num_trees_; ++tree) {
```

Comment:
Why is `this->` necessary here?

Reply:
Short answer: without `this->`, it doesn't compile (`num_trees` becomes an undefined identifier). This also applies to other instances of `this->` in this file.

Long answer: I think this has something to do with template-dependent lookup. I presume that as the base type is parameterized with `real_t`, which is a type parameter in this type, all members there become type-dependent, and are not found with a simple name lookup.

I think https://en.cppreference.com/w/cpp/language/dependent_name explains why, but I admit that I myself don't completely understand this. If you know a better way to fix this, please let me know.